StackExRedis/NHibernate.Caches.StackExRedis/RedisKeyLocker.cs

```diff
@@ -0,0 +1,244 @@
+﻿using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using NHibernate.Cache;
+using StackExchange.Redis;
+
+namespace NHibernate.Caches.StackExRedis
+{
+	/// <summary>
+	/// Provides a mechanism for locking and unlocking one or many keys.
+	/// </summary>
+	internal partial class RedisKeyLocker
+	{
+		private readonly string _regionName;
+		private readonly string _lockKeyPostfix;
+		private readonly TimeSpan _lockTimeout;
+		private readonly double _aquireLockTimeout;
+		private readonly int _retryTimes;
+		private readonly TimeSpan _maxRetryDelay;
+		private readonly TimeSpan _minRetryDelay;
+		private readonly ICacheLockRetryDelayProvider _lockRetryDelayProvider;
+		private readonly ICacheLockValueProvider _lockValueProvider;
+		private readonly IDatabase _database;
+
+		/// <summary>
+		/// Default constructor.
+		/// </summary>
+		/// <param name="regionName">The region name.</param>
+		/// <param name="database">The Redis database.</param>
+		/// <param name="configuration">The lock configuration.</param>
+		public RedisKeyLocker(
+			string regionName,
+			IDatabase database,
+			RedisCacheLockConfiguration configuration)
+		{
+			_regionName = regionName;
+			_database = database;
+			_lockKeyPostfix = configuration.KeyPostfix;
+			_lockTimeout = configuration.KeyTimeout;
+			_aquireLockTimeout = configuration.AquireTimeout.TotalMilliseconds;
+			_retryTimes = configuration.RetryTimes;
+			_maxRetryDelay = configuration.MaxRetryDelay;
+			_minRetryDelay = configuration.MinRetryDelay;
+			_lockRetryDelayProvider = configuration.RetryDelayProvider;
+			_lockValueProvider = configuration.ValueProvider;
+		}
+
+		/// <summary>
+		/// Tries to lock the given key.
+		/// </summary>
+		/// <param name="key">The key to lock.</param>
+		/// <param name="luaScript">The lua script to lock the key.</param>
+		/// <param name="extraKeys">The extra keys that will be provided to the <paramref name="luaScript"/></param>
+		/// <param name="extraValues">The extra values that will be provided to the <paramref name="luaScript"/></param>
+		/// <returns>The lock value used to lock the key.</returns>
+		/// <exception cref="CacheException">Thrown if the lock was not aquired.</exception>
+		public string Lock(string key, string luaScript, RedisKey[] extraKeys, RedisValue[] extraValues)
+		{
+			if (key == null)
+			{
+				throw new ArgumentNullException(nameof(key));
+			}
+			var lockKey = $"{key}{_lockKeyPostfix}";
+			var totalAttempts = 0;
+			var lockTimer = new Stopwatch();
+			lockTimer.Restart();
+			do
```

Comment:
This kind of retry logic would be nice to encapsulate in a class, akin to a Polly policy.

Reply:
I've added a retry policy similar to the Polly policy. The async counterpart of the policy is currently added manually as async generator does not yet support [delegate parameters](https://github.com/maca88/AsyncGenerator/issues/70).